### PR TITLE
handle `A ~ B - 1` and add tests

### LINKED ISSUE #574: Error in `Terms` when no intercept is requested
This does not work
```
df = DataFrame(A = 1:4, B = ["M", "F", "F", "M"]);
ModelMatrix(ModelFrame(A~B-1, df))
ERROR: no method convert(Type{Expr}, Array{Any,1})
 in unshift! at array.jl:511
 in Terms at /Users/andreasnoackjensen/.julia/v0.3/DataFrames/src/formula/formula.jl:145
 in ModelFrame at /Users/andreasnoackjensen/.julia/v0.3/DataFrames/src/formula/formula.jl:174
```
but this does
```
julia> ModelMatrix(ModelFrame(A~B+(-1), df))
ModelMatrix{Float64}(4x1 Array{Float64,2}:
 1.0
 0.0
 0.0
 1.0,[1])
```

### PR DESCRIPTION
**Pull request by kleinschmidt**
Fixes #574.

### CODE REVIEW

File: src/statsmodels/formula.jl [hunk 1]
```diff
@@ -40,6 +40,9 @@ type Terms
     intercept::Bool       # is there an intercept column in the model matrix?
 end
 
+import Base.==
+==(t1::Terms, t2::Terms) = all( getfield(t1, f)==getfield(t2, ) for f in fieldnames(t1))
```
Comment (nalimilan) [inline src/statsmodels/formula.jl]: We don't seem to `import` functions in this file.

Comment (kleinschmidt) [inline src/statsmodels/formula.jl]: It's not strictly necessary, just a convenience for testing `Terms`. I could move it to the test file?

Comment (nalimilan) [inline src/statsmodels/formula.jl]: I just meant that you should write `Base.:(==)(...) = ...`. Though I just noticed `f` seems to be missing in `getfield(t2, )`?

File: src/statsmodels/formula.jl [hunk 2]
```diff
@@ -85,19 +87,26 @@ function dospecials(ex::Expr)
...
-    if a1 != :* return excp end
-    aa = excp.args
-    a2 = aa[2]
-    a3 = aa[3]
-    if length(aa) > 3
-        excp.args = vcat(a1, aa[3:end])
-        a3 = dospecials(excp)
+    if a1 == :-
+        a2, a3 = excp.args[2:3]
+        a3 == 1 || error("invalid expression $ex; subtraction only supported or -1")
```
Comment (nalimilan) [inline src/statsmodels/formula.jl]: or -> for

File: test/formula.jl [hunk 3]
```diff
@@ -53,9 +53,7 @@ module TestFormula
     @test t.intercept == false
     @test t.terms == [:x1, :x2]
 
-    t = Terms(y ~ -1 + x1 + x2)
-    @test t.intercept == false
-    @test t.terms == [:x1, :x2]
+    @test t == Terms(y ~ -1 + x1 + x2) == Terms(y ~ x1 - 1 + x2) == Terms(y ~ x1 + x2 -1)
```
Comment (nalimilan) [inline test/formula.jl]: Maybe also test for the error case?

File: src/statsmodels/formula.jl (hunk 1 above)
Comment (kleinschmidt) [inline src/statsmodels/formula.jl]: That would explain the test failures :)  I tried it without the `import` first but didn't realize that `Base.:(==)` was the right way to refer to `==` in that context, thanks.

**Pull request merged**